Feature/end point

Yevgen2022 opened the pull request:
Here’s a short description of what we built:

Backend with Fastify + TypeScript.

Endpoint GET /feed with query params:

url — feed source (falls back to NASA breaking news RSS if omitted).

force — when 1/true, bypasses DB cache and parses the feed directly.

Architecture: Route → Controller → Service (functional) → Repository (Prisma).

Parsing via rss-parser, normalized to a stable DTO



{ "sourceUrl": "...", "items": [{ "title", "link", "isoDate", "pubDate", "description" }] }

Response is validated/serialized with Fastify JSON Schema.

Simple DB cache: table FeedCache (sourceUrl unique, items JSON). Without force, responses are served from cache; otherwise we fetch fresh data and upsert it.

Errors during fetch/parse return 502.

!!!!!!!!!! Current database is temporary (MySQL via Prisma). We plan to add/migrate to MongoDB as an alternative storage layer with a matching repository and a config switch.

Gunnar97 commented on src/modules/feedParser/controller/feed.controller.ts:
```diff
@@ -0,0 +1,26 @@
+	const { url, force } = request.query as {
+		url?: string;
+		force?: string | number | boolean;
+	};
```
типы квери берем из джейсон схемы

Gunnar97 commented on src/modules/feedParser/controller/feed.controller.ts:
```diff
@@ -0,0 +1,26 @@
+
+export async function getFeedDataController(
+	request: FastifyRequest,
+	reply: FastifyReply,
```
тут принимаем инстанс фастифай и логируем через него

Gunnar97 commented on src/modules/feedParser/controller/feed.controller.ts:
```diff
@@ -0,0 +1,26 @@
+	} catch (e: unknown) {
+		request.log.error(e, "feed parsing error");
+		return reply.code(502).send({ error: "Feed fetch/parse failed" });
+	}
```
у нас есть плагин sensible который позволяет делать ответы

Gunnar97 commented on src/modules/feedParser/schemas/getFeedData.schema.ts:
```diff
@@ -1,15 +1,30 @@
 	summary: "Get feed data",
-	description: "Get feed data",
 	response: {
 		200: {
```
добавить квери параметры

Gunnar97 commented on src/modules/feedParser/services/feed.service.ts:
```diff
@@ -0,0 +1,119 @@
+
+// Default feed URL
+export const DEFAULT_FEED_URL =
+	"https://www.nasa.gov/rss/dyn/breaking_news.rss";
```
я бы вынес в константы

Gunnar97 commented on src/plugins/prisma.ts:
```diff
@@ -1,22 +1,22 @@
-    const prisma = new PrismaClient()
-    await prisma.$connect()
+	const prisma = new PrismaClient();
+	await prisma.$connect();
```
тут у тебя есть подключение тогда не понятно длячего тебе либ

Yevgen2022 commented:
@Gunnar97 could you please review? CI is green
Run: npm ci && npx prisma generate && npm run dev
Test: GET /feed?url=https://news.ycombinator.com/rss&force=1
Thanks!

Gunnar97 commented:
> @Gunnar97 could you please review? CI is green Run: npm ci && npx prisma generate && npm run dev Test: GET /feed?url=https://news.ycombinator.com/rss&force=1 Thanks!

пока сервис 404 не совсем понял что я должен проверить)
<img width="831" height="989" alt="obrazek" src="https://github.com/user-attachments/assets/7fa3b639-1f57-4739-8842-60097c697b78" />

Yevgen2022 commented:
Hi  @Gunnar97 

I’ve applied all requested changes from review:

Moved Prisma logic fully into plugin (src/plugins/prisma.ts), removed lib/prisma.ts.

Repository/service accept prisma as argument, controller takes it from request.server.prisma.

Query params (url, force) are now described in JSON schema (getFeedData.schema.ts), types in controller come directly from schema (FromSchema).

force simplified to check only "1".

Errors now use reply.badGateway(...) from @fastify/sensible.

DEFAULT_FEED_URL moved to src/constants/feed.ts.

CI is green (lint, typecheck, build).

Test localy:
npm ci
npx prisma generate
npm run dev

Endpoints:
Default feed (NASA): http://localhost:3500/feed

Custom feed (Hacker News, with forced refresh):
http://localhost:3500/feed?url=https%3A%2F%2Fnews.ycombinator.com%2Frss&force=1


!!!!!!! Currently implementation uses MySQL (Prisma). MongoDB integration is not included yet, but planned in the next step.
